pkg/capi/aws.go

```diff
@@ -341,3 +338,84 @@ func newAWSMachineTemplate(name string, mapiProviderSpec *mapiv1.AWSMachineProvi
 
 	return awsMachineTemplate
 }
+
+// createCAPIMachineSetWithRetry creates a CAPI MachineSet with retry logic for capacity constraints.
+// It tries different instance types when encountering insufficient capacity errors.
+func createCAPIMachineSetWithRetry(ctx context.Context, cl client.Client, machineSetName string, clusterName string, platform configv1.PlatformType, mapiDefaultProviderSpec *mapiv1.AWSMachineProviderConfig, templateConfigurator func(*awsv1.AWSMachineTemplate, string)) (*clusterv1.MachineSet, *awsv1.AWSMachineTemplate) {
+	machineSetReady := false
+
+	// Get the current cluster architecture
+	workers, err := framework.GetWorkerMachineSets(ctx, cl)
+	Expect(err).ToNot(HaveOccurred(), "listing Worker MachineSets should not error.")
+
+	arch, err := framework.GetArchitectureFromMachineSetNodes(ctx, cl, workers[0])
+	Expect(err).NotTo(HaveOccurred(), "unable to get the architecture for the machine set")
+
+	// Select alternative instance types based on architecture
+	var alternativeInstanceTypes []string
+
+	switch arch {
+	case "arm64":
+		alternativeInstanceTypes = []string{"m6g.large", "m6g.xlarge", "c6g.large", "c6g.xlarge"}
+	default: // x86_64/amd64
+		alternativeInstanceTypes = []string{"m5.large", "m5.xlarge", "c5.xlarge", "m5a.large"}
```

Comment:
Are we now excluding the m6 category on purpose?

Reply:
No, I picked some randomly and I don't think this case is related to the instance type. If you want to test m6 category, I can also update it.